### ISSUE TITLE
Re-initialize map view after changing style 

### ISSUE THREAD

**Issue by zstadler** [maplibre/maputnik#871]
**Maputnik version**: main
**Browser**: Chrome, Firefox
**OS**: Windows

**Description of the bug**:
When changing style, the map view may contain features of the previous style

**Steps to reproduce the behavior**:
1. Open Maplibre's [maputnik in this location](https://maplibre.org/maputnik/#13/31.27563/35.12306)
2. Open -> Gallery Styles -> Empty Style
3. Reload the page to be on the safe side
4. Open -> Open URL -> https://raw.githubusercontent.com/IsraelHikingMap/VectorMap/master/Styles/IHM.json
5. The map view would look like this:
   ![image](https://github.com/maplibre/maputnik/assets/1304610/97b37048-a867-4be4-8247-a37d0387c698)
6. Open -> Gallery Styles -> OSM Liberty
7. The map view would look like this, with a mix of both styles:
   ![image](https://github.com/maplibre/maputnik/assets/1304610/99a6af70-446c-4a68-807d-a42d94879e09)
8. Reload the page
9. The map view would look like this, without traces of the previous style:
   ![image](https://github.com/maplibre/maputnik/assets/1304610/9fb39065-1921-43ed-be4a-760ed8efcd65)

**Comment by HarelM** [comment]
Hmm... interesting, might be related to style diff capabilities or limitations.

**Comment by zstadler** [comment]
FWIW, cannot reproduce this with v1.7.0 at https://maputnik.github.io/editor/#13/31.27563/35.12306

**Comment by HarelM** [comment]
Looks like there's wrong handling of style update somewhere in the code.
The following is reported in the console:
![image](https://github.com/maplibre/maputnik/assets/3269297/b63e898a-386a-4b97-b764-67e60ada7a79)

In the past (1.7), the style diff was not able to perform the diff and the map was reloaded with the following warning:
`Unable to perform style diff: Unimplemented: setSprite..  Rebuilding the style from scratch.`
In recent version of maplibre-gl-js there is a more eager attempt to allow style diffing, but it might have bugs.
So this might be a bug in maplibre and not maputnik.
I'll need to see if I can create a minimal reproduction.
Thanks for reporting this issue!

**Comment by HarelM** [comment]
Ok, so I tried to see if I can reproduce this on "plain" maplibre, and I was not able to.
I also looked at the diff code in both maplibre-gl and maplibre-style-spec to see if I see some problems there and it is both covered with tests and runs as expected.
So I dug deeper and it seems that the reason for this is the following:
1. When replacing a style, the style diff method is called.
2. This in turn removes all the layers and then removes the sources.
3. After the first source is removed, an event of `data` is raised to let anyone who listens know that the source was removed.
4. This apparently is picked up by mapbox-inspect which forces the original style back to the map.

mapbox-inspect is updating the map style here:
https://github.com/lukasmartinelli/mapbox-gl-inspect/blob/dab82f2f83670bfeea93195da5db0f35dae99921/lib/MapboxInspect.js#L130

It might make sense since mapbox-inspect has its own style, but I'm not sure it should be applied when the mode in not "inspect" mode.

Bottom line, this is a maputnik related, I'm not sure I fully understand how to solve it though...

**Comment by nyurik** [comment]
We could ask @lukasmartinelli if he would be interested in transferring it here too - seems like a worthy component to maintain.

**Comment by lukasmartinelli** [comment]
Happy to transfer it!

**Comment by nyurik** [comment]
Awesome, thanks @lukasmartinelli!  I think the easiest is to grant me or @HarelM admin rights on that repo, and we can do the rest.  Note that we will probably want to continue publishing it under the original npm name, so may need to add harel as an admin there too. Thx!

**Comment by nyurik** [comment]
Actually, I just realized it has `mapbox` in its name - so perhaps we should rename it and/or fork it, and republish under maplibre name?

**Comment by HarelM** [comment]
Yeah, I was thinking about `@maplibre/maplibre-inspect` or something similar like we did with the other libraries.
We can keep the original repo as is for legacy mapbox users.

**Comment by HarelM** [comment]
I can probably commit to maintaining it, as it is an integral part of maputnik, and I don't expect a lot of changes there.
@nyurik will you do the honor of opening a PR to fork the repo under the maplibre umbrella?

**Comment by nyurik** [comment]
Yeah, perhaps it really does make sense to have an official fork that does not show up as a "github fork" (under the name at the top), but instead mentions in the readme that it was adapted from the other repo and renamed.  There are not that many issues/pull requests it seems to do the transfer...

**Comment by nyurik** [comment]
I created https://github.com/maplibre/maplibre/issues/359 - I guess we should do a proper fork (without the link to upstream).  Leaving it to @HarelM :)

**Comment by HarelM** [comment]
@zstadler let me know if you can reproduce this issue in latest version, I have deployed a temporary fix until we migrate maplibre-gl-inspect under maplibre org.

**Comment by zstadler** [comment]
@HarelM,
The original steps to reproductions do not trigger this issue.

However, since inspect was mentioned above, I've tried the following steps, and the style does change as expected.

1. Open Maplibre's [maputnik in this location](https://maplibre.org/maputnik/#13/31.27563/35.12306)
2. Open -> Gallery Styles -> Empty Style
3. Reload the page to be on the safe side
4. Open -> Open URL -> https://raw.githubusercontent.com/IsraelHikingMap/VectorMap/master/Styles/IHM.json
5. **_Change to inspect view_**
6. Open -> Gallery Styles -> OSM Liberty
7. **_Change to map view_**
8. The Israel Hiking style is shown rather than OSM Liberty
9. Reload the page
10. The OSM Liberty style map is shown

**Comment by HarelM** [comment]
The reason for the above, which is similar but different to the original issue is that the inspect library adds metadata to the style to indicate that this style is from the inspect library.
When doing the diff, this flag is kept and the inspect library doesn't change the "original" style.
So when turning off inspect, the original style is incorrect.
Seems like there are a lot of assumptions in both maputnik and inspect that are preventing this from working correctly, I'm not sure I'll be able to solve those...

**Comment by HarelM** [comment]
After an archaeological digging I found some hacks that were added in order to fix some odd behavior between mapbox-gl and mapbox-gl-inspect.
Here is the relevant link: https://github.com/maplibre/maputnik/blob/5f54dd0ccf34d205c8598186bd41eaae328f238b/src/components/MapMaplibreGl.tsx#L120L132

We now own maplibre-inspect and these should be fixed properly now in there, the problem is the issues they link to, and making sure everything is working as expected after the hacks are removed.
In general there are some race conditions between setting the inspect style and regular style, and it causes a mayham.

Relevant issues to the above HACK comments:
https://github.com/maplibre/maputnik/issues/576
https://github.com/maplibre/maputnik/issues/647

I'm not sure I know how to solve all this "properly" as it seems there's a fundamental problem with how maputnik works with the store and event drivern and how maplibre-inspect works - listening to map events.

**Comment by HarelM** [comment]
I hope both issues reported here are fixed now.
I have tested both on the local dev machine and now the deployed site I can't reproduce the issues.
Due to the complexity of the events and sync, the solution caused the transition to and from inspect is a bit more sluggish, but I'm OK with that assuming the data is correct.

**Comment by zstadler** [comment]
Well, almost...

The second scenario is leaves Maputnik in an inconsistent state

1. Open Maplibre's [maputnik in this location](https://maplibre.org/maputnik/#13/31.27563/35.12306)
2. Open -> Gallery Styles -> Empty Style
3. Reload the page to be on the safe side
4. Open -> Gallery Styles -> OSM Liberty
5. **_Change to inspect view_**
6. Open -> Gallery Styles -> Positron
8. The Positron style map is shown rather inspect view
  ![image](https://github.com/maplibre/maputnik/assets/1304610/3abe29f2-d7fd-4d44-a0f3-b79f016c8b6b)

**Comment by HarelM** [comment]
In the above case, the sources are equal, so after the style change, which sets the "regular" map style, the inspect code, which listens to style changes compares the sources and doesn't apply the inspect style.
I'll check what can be done, thanks for all the testing!!

**Comment by HarelM** [comment]
Hopefully the last patch I'm doing to fix these issues: #889

**Comment by zstadler** [comment]
LGTM!
